tests/unittest_qmail-send.c

```diff
@@ -0,0 +1,239 @@
+  numjobs = -77;
+}
+
+static void teardown(void) {
+  numjobs = -88;
+}
+
+START_TEST(test_qmail_job_init_sets_expected_default_state)
+{
+  ck_assert_ptr_null(jo);
+
+  job_init_this_many(4);
+
+  ck_assert_ptr_nonnull(jo);
+  ck_assert_uint_eq(numjobs, 4);
+  assert_job_at_index(0, available);
+  ck_assert_ptr_eq(jo[0].sender.s, 0);
+  assert_job_at_index(3, available);
+  ck_assert_ptr_eq(jo[3].sender.s, 0);
+}
+END_TEST
+
+START_TEST(test_qmail_job_avail_returns_free_slots_after_init)
+{
+  job_init_this_many(5);
+  assert_job_avail(job_avail(), 1);
+}
+END_TEST
+
+START_TEST(test_qmail_job_avail_returns_no_free_slots)
+{
+  job_init_this_many(5);
+  occupy_all_jobs();
+
+  assert_job_avail(job_avail(), 0);
+}
+END_TEST
+
+START_TEST(test_qmail_job_avail_returns_a_free_slot)
+{
+  job_init_this_many(5);
+  occupy_all_jobs();
+  release_job(2);
+
+  assert_job_avail(job_avail(), 1);
+}
+END_TEST
+
+START_TEST(test_qmail_job_avail_returns_a_free_slot_for_index_zero)
+{
+  job_init_this_many(5);
+  occupy_all_jobs();
+  release_job(0);
+
+  assert_job_avail(job_avail(), 1);
+}
+END_TEST
+
+START_TEST(test_qmail_job_avail_returns_a_free_slot_for_last_index)
+{
+  job_init_this_many(5);
+  occupy_all_jobs();
+  release_job(numjobs - 1);
+
+  assert_job_avail(job_avail(), 1);
+}
+END_TEST
+
+START_TEST(test_qmail_job_open_slots_full)
+{
+  job_init_this_many(2);
+  occupy_all_jobs();
+
+  int index = job_open(1,2);
+
+  ck_assert_int_eq(index, -1);
+}
+END_TEST
+
+START_TEST(test_qmail_job_open_slot_available)
+{
+  job_init_this_many(5);
+  occupy_all_jobs();
+  release_job(3);
+  unsigned long job_open_id = 6546;
+  int job_open_channel = 5;
+
+  int index = job_open(job_open_id,job_open_channel);
+
+  ck_assert_int_eq(index, 3);
+  assert_job_at_index(3, unavailable);
+  ck_assert_uint_eq(jo[3].id, job_open_id);
+  ck_assert_int_eq(jo[3].channel, job_open_channel);
+  ck_assert_int_eq(jo[3].numtodo, 0);
+  ck_assert_int_eq(jo[3].flaghiteof, 0);
+}
+END_TEST
+
+START_TEST(test_qmail_job_open_multiple_slots_available)
+{
+  job_init_this_many(7);
+  occupy_all_jobs();
+  release_job(2);
+  release_job(3);
+  unsigned long job_open_id = 7890;
+  int job_open_channel = 12;
+
+  int index = job_open(job_open_id,job_open_channel);
+
+  ck_assert_int_eq(index, 2);
+  assert_job_at_index(2, unavailable);
+  ck_assert_uint_eq(jo[2].id, job_open_id);
+  ck_assert_int_eq(jo[2].channel, job_open_channel);
+  ck_assert_int_eq(jo[2].numtodo, 0);
+  ck_assert_int_eq(jo[2].flaghiteof, 0);
+}
+END_TEST
+
+/*
+ * job_close() could be called multiple times, dropping refs below 0.
+ * It doesn't look like qmail-send ever does this.
+ * But if we understand refs, we'd prefer boolean: available XOR unavailable.
+ */
+START_TEST(test_qmail_job_close_decrements_available_slots_for_each_call)
+{
+  job_init_this_many(7);
+  assert_job_at_index(3, available);
+
+  job_close(3);
+  assert_job_at_index(3, -1);
+  job_close(3);
+  assert_job_at_index(3, -2);
+}
+END_TEST
```

Comment:
I'm unsure about this test. It does test the current behavior, but as your comment shows you also see this as questionable. I would not test for this behavior as we may actually change it to something different, and suddenly having a failing test then could lead somebody to think that this behavior is actually wanted.

Reply:
Thanks for the commentary here @DerDakon 

Wouldn't a subsequent PR that removes the behavior also include the rewritten, or newly written test?

I agree that having suddenly failing test could lead somebody to look into the test, but I also think that at some point this test would be updated to assert that the value will never be below zero.

I think that this characterization test is pretty useful since we can always see with certainty what the program is doing today, and we can determine if it should change and be more explicit about it's behavior in the future, for example, if available-slot (refs) is already set to 0 value (available), do not decrement, or explicitly set the value to 0, instead of a decrementing the value of refs in the function call.

Reply:
This is a counter overflow, and I can't think that it was intentional. When you do it like this you will run in odd behavior at other places, I have tested this while debugging another thing. The implicit assumption seems to be that this never underflows, which it doesn't if you do everything correct. You can see that other places e.g. check for ```refs != 0``` or ```!refs```, this was never intended for having anything negative.